flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraph.java

```diff
@@ -1215,7 +1215,19 @@ private void jobFinished() {
 
             try {
                 for (ExecutionJobVertex ejv : verticesInCreationOrder) {
-                    ejv.getJobVertex().finalizeOnMaster(getUserClassLoader());
+                    ejv.getJobVertex()
+                            .finalizeOnMaster(
+                                    new JobVertex.InitializeOnMasterContext() {
+                                        @Override
+                                        public ClassLoader getClassLoader() {
+                                            return DefaultExecutionGraph.this.getUserClassLoader();
+                                        }
+
+                                        @Override
+                                        public int getExecutionParallelism() {
+                                            return ejv.getParallelism();
+                                        }
+                                    });
```

Comment:
Let's add a simple pojo for this instead of generating a new class for each vertex. Should be more readable (and more performant I suppose).